### ISSUE TITLE
[Bug] LazyEvaluatedKernelTensor does not require gradient if inputs do not require gradient

### ISSUE THREAD

**Issue by wjmaddox** [cornellius-gp/gpytorch#1511]
# 🐛 Bug

Not entirely sure if this is intended behavior, but it does produce some downstream problems. 
If the input to `kernel(x)` doesn't require a gradient `LazyEvaluatedKernelTensor` does not portray as requiring a gradient, even though it does. Crucially `kernel(x).evaluate_kernel()` does. 

As a result, it causes some weirdness downstream in objects that can manipulate lazy kernel tensors themselves:

## To reproduce

** Code snippet to reproduce **
```python
import torch
from gpytorch.kernels import MaternKernel
from gpytorch.lazy import KroneckerProductLazyTensor

torch.random.manual_seed(3)

kernels = [MaternKernel()] * 3
x = [torch.randn(5)] * 3

kernels[0](x[0]).requires_grad # False

# breaking behavior
kplt = KroneckerProductLazyTensor(*[k(a) for k, a in zip(kernels, x)])

rhs = torch.randn(kplt.shape[-1])
print(kplt.matmul(rhs).requires_grad) # True, as expected

kdiag = kplt.diag()
print(kdiag.requires_grad) # False ??

# the fix is that this works fine
kplt = KroneckerProductLazyTensor(*[k(a).evaluate_kernel() for k, a in zip(kernels, x)])
kdiag = kplt.diag()
print(kdiag.requires_grad) # True
```

## Expected Behavior

LazyKernelEvaluatedTensor should probably require a gradient to prevent this sort of behavior from happening.

## System information

**Please complete the following information:**
gpytorch master

## Additional context

I'm trying to implement structured variational MTGP models with non-gaussian likelihoods and am struggling to get reasonable gradients, noticed while debugging.

**Comment by gpleiss** [comment]
```python
# breaking behavior
kplt = KroneckerProductLazyTensor(*[k(a) for k, a in zip(kernels, x)])

kdiag = kplt.diag()
print(kdiag.requires_grad) # False ??
```

This is actually correct behavior for this specific script: when calling `.diag` of a stationary kernel, we populate the distance function with a zeros vector (which doesn't require grad). See https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/kernels/kernel.py#L312

However, in the more general case (e.g. replace `MaternKernel` with `LinearKernel` in your example), then yes there is breaking behavior. I'm putting up a PR to fix it.
